src/AutorestSwift/Models/Complex/ParameterType.swift

```diff
@@ -0,0 +1,212 @@
+            return virt as Parameter
+        }
+    }
+}
+
+extension ParameterType: Equatable {
+    static func == (lhs: ParameterType, rhs: ParameterType) -> Bool {
+        switch lhs {
+        case let .regular(lparam):
+            if case let ParameterType.regular(rparam) = rhs {
+                return lparam == rparam
+            }
+        case let .virtual(lparam):
+            if case let ParameterType.virtual(rparam) = rhs {
+                return lparam == rparam
+            }
+        }
+        return false
+    }
+}
+
+extension Array where Element == ParameterType {
+    func first(named: String) -> Element? {
+        for param in self {
+            let name = param.serializedName ?? param.name
+            if named == name {
+                return param
+            }
+        }
+        // no match found
+        return nil
+    }
+
+    /// Returns the subset of `ParameterType` that are `VirtualParameter` types.
+    var virtual: [VirtualParameter] {
+        return compactMap { param in
+            if case let ParameterType.virtual(virtParam) = param {
+                return virtParam
+            }
+            return nil
+        }
+    }
+
+    /// Returns the required items that should be in the method signature
+    var inSignature: [ParameterType] {
+        return filter { param in
+            guard param.implementation == ImplementationLocation.method,
+                param.schema.type != .constant
+            else { return false }
+            // We track body params in a special way, so always omit them generally from the signature.
+            // If they belong in the signature, they will be added in a way to ensure they come first.
+            if let httpParam = param.protocol.http as? HttpParameter {
+                guard httpParam.in != .body else { return false }
+            }
+            return param.required
+        }
+    }
+
+    /// Returns the optional items that should be in the method's Options object
+    var inOptions: [ParameterType] {
+        return filter { param in
+            guard param.implementation == ImplementationLocation.method,
+                param.schema.type != .constant,
+                param.flattened == false
+            else { return false }
+            return !param.required
+        }
+    }
+
+    /// Returns the items that should be passed in the query string
+    var inQuery: [ParameterType] {
+        return filter { param in
+            guard param.implementation == ImplementationLocation.method,
+                param.schema.type != .constant
+            else { return false }
+            if let httpParam = param.protocol.http as? HttpParameter {
+                return httpParam.in == .query
+            } else {
+                return false
+            }
+        }
+    }
+
+    /// Returns the items that should be passed in the request header
+    var inHeader: [ParameterType] {
+        return filter { param in
```

Comment:
the method inHeader, InPath, inSignature have. a commonitity. May be create a helper function and take what to check in httpParam.in for each function?

Reply:
Done.